p2p/src/sync/peer_v2/pending_transactions.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{cmp::Reverse, collections::BinaryHeap, time::Duration};
+
+use common::{
+    chain::Transaction,
+    primitives::{Id, H256},
+    time_getter::TimeGetter,
+};
+
+use crate::Result;
+
+pub struct PendingTransactions {
+    time_getter: TimeGetter,
+    txs: BinaryHeap<Reverse<(Duration, Id<Transaction>)>>,
+}
+
+impl PendingTransactions {
+    pub fn new(time_getter: TimeGetter) -> Self {
+        Self {
+            time_getter,
+            txs: Default::default(),
+        }
+    }
+
+    pub fn push(&mut self, tx: Id<Transaction>, due_time: Duration) {
+        self.txs.push(Reverse((due_time, tx)));
+    }
+
+    pub async fn due(&mut self) -> Result<Id<Transaction>> {
```

Comment:
I think it's better if `due` did not mutate the container to make the future cancellation safe [1]. A separate method will be needed to take the top entry. That way it will behave correctly if the future object is dropped. It may happen for example in `select!` with pattern match. I think it's fine the way it is used now but preferably the user does not need to think about it, making the interface simpler to reason about.

The downside of having a separate method is that the future checks if the value is there and you immediately pull it out using the other method. That introduces the Rust anti-pattern of having checks and return value separate instead of returning something like `Option`. That could be solved by returning some sort of `TopEntry` struct that allows you to pull out the top element later but that's maybe too convoluted?

Alternatively, just think through all the cancellation scenarios and write a big fat comment explaining why it's fine and under which conditions can this method be used without losing data.

By the way this returns `Result` but I don't see any errors.

[1] https://docs.rs/tokio/1.35.0/tokio/macro.select.html#cancellation-safety

Reply:
Agree, though it was cancellation safe I split the method anyway